staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go

```diff
@@ -309,6 +334,10 @@ func (r *RuntimeSort) Less(i, j int) bool {
 	iField := iValues[0][0]
 	jField := jValues[0][0]
 
+	// Convert Quantity to string for test cases
```

Comment:
Can you please help me understand why we need to do this?

Reply:
Thanks for the explanation. We should avoid this dance if possible. Might need to change how we're doing the reflection. Need someone that knows Go way better than I to take a look.

@KnVerey @seans3 can either of you please take a look?

Reply:
Specific failure without this:

```
    --- FAIL: TestSortingPrinter/pod-resources-cpu-random-order-with-missing-fields (0.00s)
panic: reflect.Value.Interface: cannot return value obtained from unexported field or method [recovered]
        panic: reflect.Value.Interface: cannot return value obtained from unexported field or method

goroutine 78 [running]:
testing.tRunner.func1.2(0x5521bc0, 0x5a08620)
        /usr/local/Cellar/go/1.16.3/libexec/src/testing/testing.go:1143 +0x332
testing.tRunner.func1(0xc000783200)
        /usr/local/Cellar/go/1.16.3/libexec/src/testing/testing.go:1146 +0x4b6
panic(0x5521bc0, 0x5a08620)
        /usr/local/Cellar/go/1.16.3/libexec/src/runtime/panic.go:965 +0x1b9
reflect.valueInterface(0x5721c80, 0xc00070a280, 0xb9, 0xc0006a3201, 0x58ca738, 0xc0006b77a8)
        /usr/local/Cellar/go/1.16.3/libexec/src/reflect/value.go:1027 +0x169
reflect.Value.Interface(...)
        /usr/local/Cellar/go/1.16.3/libexec/src/reflect/value.go:1016
k8s.io/kubectl/pkg/cmd/get.isLess(0x5721c80, 0xc00070a280, 0xb9, 0x5721c80, 0xc00070a2c0, 0xb9, 0xb9, 0x6f23108, 0x18)
        /Users/ezaneski/Codez/kubernetes/kubernetes/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:200 +0x15dd
k8s.io/kubectl/pkg/cmd/get.isLess(0x5728f40, 0xc00070a280, 0x99, 0x5728f40, 0xc00070a2c0, 0x99, 0x0, 0x0, 0xc0006a30b0)
        /Users/ezaneski/Codez/kubernetes/kubernetes/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:207 +0x1750
k8s.io/kubectl/pkg/cmd/get.(*RuntimeSort).Less(0xc00070a180, 0x1, 0x0, 0x30)
        /Users/ezaneski/Codez/kubernetes/kubernetes/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:341 +0x55a
sort.insertionSort(0x5a58530, 0xc00070a180, 0x0, 0x4)
        /usr/local/Cellar/go/1.16.3/libexec/src/sort/sort.go:40 +0xc4
sort.quickSort(0x5a58530, 0xc00070a180, 0x0, 0x4, 0x6)
        /usr/local/Cellar/go/1.16.3/libexec/src/sort/sort.go:222 +0x205
sort.Sort(0x5a58530, 0xc00070a180)
        /usr/local/Cellar/go/1.16.3/libexec/src/sort/sort.go:231 +0x76
k8s.io/kubectl/pkg/cmd/get.SortObjects(0xef2c918, 0xc0005b6e60, 0xc00070a040, 0x4, 0x4, 0x582f039, 0x2b, 0xc0005b6e60, 0xc0001e9110, 0x5a25400)
        /Users/ezaneski/Codez/kubernetes/kubernetes/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:146 +0x45d
k8s.io/kubectl/pkg/cmd/get.(*SortingPrinter).sortObj(0xc00008af10, 0x5a2e680, 0xc0000e08c0, 0xc00051d680, 0x3)
        /Users/ezaneski/Codez/kubernetes/kubernetes/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go:83 +0xb9
k8s.io/kubectl/pkg/cmd/get.TestSortingPrinter.func3(0xc000783200)
        /Users/ezaneski/Codez/kubernetes/kubernetes/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter_test.go:601 +0xf9
testing.tRunner(0xc000783200, 0xc000406b40)
        /usr/local/Cellar/go/1.16.3/libexec/src/testing/testing.go:1193 +0xef
created by testing.(*T).Run
        /usr/local/Cellar/go/1.16.3/libexec/src/testing/testing.go:1238 +0x2b3
FAIL    k8s.io/kubectl/pkg/cmd/get      0.615s
FAIL
```

Reply:
I agree with @eddiezane - we should not need to do that. It seems to me that what's happening is that the test data is very differently typed from the input the command would actual deal with. Specifically, in this case the test passes in corev1.Pod objects, whereas the "real thing" deals in unstructured.Unstructured. So in the test code path, the jsonpath `.spec.containers[0].resources.requests.cpu` resolves to a `resource.Quantity` struct, so we hit [this case](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go#L183). We then [iterate over the fields](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go#L191-L196) looking at their types, which blows up because they're (mostly) private. 

So I think the correct way to fix this _for the tests_ (and any other callers that might pass in the real typed structs, since this package is public) is to add `resource.Quantity` handling right there, the same way the code already does for [`metav1.Time`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/get/sorter.go#L186-L189):

```go
		if iQuantity, ok := in.(resource.Quantity); ok {
			jQuantity := j.Interface().(resource.Quantity)
			return iQuantity.Cmp(jQuantity) < 0, nil
		}
```

We probably want test coverage for the "real" code path too though! We actually kinda already have some--remove the string-type changes and the table tests fail, because their setup converts the input to unstructured.Unstructured. That said, we could explicitly write tests this way too, e.g.:

```go
func createUnstructuredPodResource(memReq, memLimit, cpuReq, cpuLimit string) unstructured.Unstructured {
	pod := &corev1.Pod{
		Spec: createPodSpecResource(memReq, memLimit, cpuReq, cpuLimit),
	}
	return *toUnstructuredOrDie(encodeOrDie(pod))
}

{
			name: "pod-unstructured-resources-cpu-random-order-with-missing-fields",
			obj: &unstructured.UnstructuredList{
				Object: map[string]interface{}{
					"kind":       "List",
					"apiVersion": "v1",
				},
				Items: []unstructured.Unstructured{
					createUnstructuredPodResource("", "", "0.5", ""),
					createUnstructuredPodResource("", "", "10", ""),
					createUnstructuredPodResource("", "", "100m", ""),
					createUnstructuredPodResource("", "", "", ""),
				},
			},
			sort: &unstructured.UnstructuredList{
				Object: map[string]interface{}{
					"kind":       "List",
					"apiVersion": "v1",
				},
				Items: []unstructured.Unstructured{
					createUnstructuredPodResource("", "", "", ""),
					createUnstructuredPodResource("", "", "100m", ""),
					createUnstructuredPodResource("", "", "0.5", ""),
					createUnstructuredPodResource("", "", "10", ""),
				},
			},
			field: "{.spec.containers[].resources.requests.cpu}",
		},
```

Wdyt? Would that be overkill?

Regardless, **I think we should have integration test coverage for this change**. Luckily, it seems to be pretty simple to add:

1. Add resource fields to [these test pods](https://github.com/kubernetes/kubernetes/tree/master/hack/testdata/sorted-pods)
2. Add a new assertion [somewhere around here](https://github.com/kubernetes/kubernetes/blob/master/test/cmd/get.sh#L317) e.g.:

```bash
 # ensure sorting by resource quantity works
  output_message=$(kubectl get pods --sort-by="{.spec.containers[0].resources.requests.cpu}")
  kube::test::if_sort_by_has_correct_order "${output_message}" "sorted-pod3:sorted-pod1:sorted-pod2:"
```

Reply:
will take a deeper look soon

Reply:
@KnVerey I actually thought of adding this, 
```
if iQuantity, ok := in.(resource.Quantity); ok {
			jQuantity := j.Interface().(resource.Quantity)
			return iQuantity.Cmp(jQuantity) < 0, nil
		}
```
but it won't reflect what happens when we want to sort a `Resource` because the `Resource` will be treated as `string` instead of `Resource` as explained in `kubectl explain pod.spec.containers.resources`
```
KIND:     Pod
VERSION:  v1

RESOURCE: resources <Object>

DESCRIPTION:
     Compute Resources required by this container. Cannot be updated. More info:
     https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/

     ResourceRequirements describes the compute resource requirements.

FIELDS:
   limits	<map[string]string>
     Limits describes the maximum amount of compute resources allowed. More
     info:
     https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/

   requests	<map[string]string>
     Requests describes the minimum amount of compute resources required. If
     Requests is omitted for a container, it defaults to Limits if that is
     explicitly specified, otherwise to an implementation-defined value. More
     info:
     https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/

```
so I don't think changing to `unstructured.Unstructured` type is overkill

Reply:
I agree that the `resource.Quantity` case would not get hit when invoking `kubectl`, which is the case where the input will end up being an `unstructured.Unstructued`. The reason we might want to add that case anyway is that this code can be used in other contexts and can in fact accept specifically typed input, as your initial tests cases showed. If we don't add the `resource.Quantity` handling (similar to the Time handling already present), sorting will continue to be incorrect when the code is called directly with e.g. a `v1.Pod`. Perhaps an approver can clarify whether or not that matters.

Reply:
Sure I will keep that then 👍🏼